### PR TITLE
V6 replace uni v2 router and add polygon

### PR DESCRIPTION
**Pull request by charlesjhongc**
- Replace old Uniswap v2 router with `SwapRouter02`
- Add polygon network config

### CODE REVIEW

File: test/utils/UniswapV2Library.sol [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+        uint256 denominator = (reserveIn * 1000) + amountInWithFee;
+        amountOut = numerator / denominator;
+    }
+
+    // performs chained getAmountOut calculations on any number of pairs
+    function getAmountsOut(uint256 amountIn, address[] memory path) internal view returns (uint256[] memory amounts) {
+        require(path.length >= 2, "UniswapV2Library: INVALID_PATH");
+        amounts = new uint256[](path.length);
+        amounts[0] = amountIn;
+        for (uint256 i; i < path.length - 1; i++) {
```
Comment (alex0207s) [inline test/utils/UniswapV2Library.sol:119]: Replace the suffix increment with the prefix one.

File: test/utils/UniswapV2Library.sol [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+        require(amountIn > 0, "UniswapV2Library: INSUFFICIENT_INPUT_AMOUNT");
+        require(reserveIn > 0 && reserveOut > 0, "UniswapV2Library: INSUFFICIENT_LIQUIDITY");
+        uint256 amountInWithFee = amountIn * 997;
+        uint256 numerator = amountInWithFee * reserveOut;
+        uint256 denominator = (reserveIn * 1000) + amountInWithFee;
+        amountOut = numerator / denominator;
+    }
+
+    // performs chained getAmountOut calculations on any number of pairs
+    function getAmountsOut(uint256 amountIn, address[] memory path) internal view returns (uint256[] memory amounts) {
```
Comment (alex0207s) [inline test/utils/UniswapV2Library.sol:115]: Use calldata to store the path is better here.

File: test/utils/config/arbitrumMainnet.json [hunk 3]
```diff
@@ -7,31 +7,11 @@
...
-  "cUSDC_ADDRESS": "0x0000000000000000000000000000000000000000",
-  "yDAI_ADDRESS": "0x0000000000000000000000000000000000000000",
-  "yUSDC_ADDRESS": "0x0000000000000000000000000000000000000000",
-  "yUSDT_ADDRESS": "0x0000000000000000000000000000000000000000",
-  "yTUSD_ADDRESS": "0x0000000000000000000000000000000000000000",
-
-  "ARBITRUM_L1_GATEWAY_ROUTER_ADDR": "0x0000000000000000000000000000000000000000",
-  "ARBITRUM_L1_BRIDGE_ADDR": "0x0000000000000000000000000000000000000000",
-  "OPTIMISM_L1_STANDARD_BRIDGE_ADDR": "0x0000000000000000000000000000000000000000"
+  "UNISWAP_UNIVERSAL_ROUTER_ADDRESS": "0x4648a43B2C14Da09FdF82B161150d3F634f40491"
```
Comment (108356037) [inline test/utils/config/arbitrumMainnet.json:16]:
```suggestion
  "UNISWAP_UNIVERSAL_ROUTER_ADDRESS": "0x4C60051384bd2d3C01bfc845Cf5F4b44bcbE9de5"
  "UNISWAP_UNIVERSAL_ROUTER_ADDRESS_V1_2": "0xeC8B0F7Ffe3ae75d7FfAb09429e3675bb63503e4"
```
Addresses should be these according to latest code: https://github.com/Uniswap/universal-router/blob/main/deploy-addresses/arbitrum.json

**Review by alex0207s — COMMENTED** [review]
Leave some comments for the better practices, overall LGTM.

File: test/utils/config/polygon.json [hunk 4]
```diff
@@ -0,0 +1,17 @@
+{
+  "WETH_ADDRESS": "0x0d500B1d8E8eF31E21C99d1Db9A6444d3ADf1270",
+  "USDT_ADDRESS": "0xc2132D05D31c914a87C6611C10748AEb04B58e8F",
+  "USDC_ADDRESS": "0x2791Bca1f2de4661ED88A30C99A7a9449Aa84174",
+  "CRV_ADDRESS": "0x172370d5Cd63279eFa6d502DAB29171933a610AF",
+  "TUSD_ADDRESS": "0x2e1AD108fF1D8C782fcBbB89AAd783aC49586756",
+  "DAI_ADDRESS": "0x8f3Cf7ad23Cd3CaDbD9735AFf958023239c6A063",
+  "LON_ADDRESS": "0x6f7C932e7684666C9fd1d44527765433e01fF61d",
```
Comment (108356037) [inline test/utils/config/polygon.json:8]:
```suggestion
  "MKR_ADDRESS": "0x6f7C932e7684666C9fd1d44527765433e01fF61d",
  "LON_ADDRESS": "0x000000000000000000000000000000000000000",
```
This address belongs to MAKER on Polygon: https://polygonscan.com/address/0x6f7C932e7684666C9fd1d44527765433e01fF61d

File: contracts/AMMStrategy.sol
Comment (alex0207s) [inline contracts/AMMStrategy.sol:1]: This strategy is for testing purposes, there is no need to keep it.

File: contracts/UniswapStrategy.sol
Comment (alex0207s) [inline contracts/UniswapStrategy.sol:1]: This strategy is for testing purposes, there is no need to keep it.

File: test/utils/config/polygon.json [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+  "WBTC_ADDRESS": "0x1BFD67037B42Cf73acF2047067bd4F2C47D9BfD6",
+
+  "CURVE_TRICRYPTO2_POOL_ADDRESS": "0x000000000000000000000000000000000000000",
+  "SUSHISWAP_ADDRESS": "0x000000000000000000000000000000000000000",
+  "UNISWAP_V3_QUOTER_ADDRESS": "0xb27308f9F90D607463bb33eA1BeBb41C27CE5AB6",
+  "UNISWAP_PERMIT2_ADDRESS": "0x000000000022d473030f116ddee9f6b43ac78ba3",
+  "UNISWAP_SWAP_ROUTER_02_ADDRESS": "0x68b3465833fb72A70ecDF485E0e4C7bD8665Fc45",
+  "UNISWAP_UNIVERSAL_ROUTER_ADDRESS": "0x3fC91A3afd70395Cd496C647d5a6CC9D4B2b7FAD"
```
Comment (108356037) [inline test/utils/config/polygon.json:16]:
```suggestion
  "UNISWAP_UNIVERSAL_ROUTER_ADDRESS": "0x4C60051384bd2d3C01bfc845Cf5F4b44bcbE9de5"
```
Universal router is `0x4C60051384bd2d3C01bfc845Cf5F4b44bcbE9de5` according to latest code: https://github.com/Uniswap/universal-router/blob/main/deploy-addresses/polygon.json
However `0x3fC91A3afd70395Cd496C647d5a6CC9D4B2b7FAD` is still a workable address at the moment.

File: test/utils/config/polygon.json (hunk 4 above)
Comment (charlesjhongc) [reply test/utils/config/polygon.json:8]: Yeah but we don't have LON on polygon.

File: test/utils/UniswapV2Library.sol (hunk 1 above)
Comment (charlesjhongc) [reply test/utils/UniswapV2Library.sol]: Fixed in 3652669b684d44a95d8b5530fbfa71fe58c22b79

File: test/utils/UniswapV2Library.sol (hunk 2 above)
Comment (charlesjhongc) [reply test/utils/UniswapV2Library.sol:115]: We create path in memory in tests so can't use calldata here. Also, these utils are for test only so we won't spend too much effort on optimizing here.

File: test/utils/config/arbitrumMainnet.json (hunk 3 above)
Comment (charlesjhongc) [reply test/utils/config/arbitrumMainnet.json]: Fixed 22e1a5928e739ca5c844ca10b7f7ea745bafb9a8

File: test/utils/config/polygon.json (hunk 5 above)
Comment (charlesjhongc) [reply test/utils/config/polygon.json]: Fixed 22e1a5928e739ca5c844ca10b7f7ea745bafb9a8

**Comment by charlesjhongc** [comment]
Thank you @108356037 for carefully checking the addresses. I found that Uniswap no longer use same address across multiple networks so can't assume it's all the same anymore.

**Review by 108356037 — APPROVED** [review]

File: test/utils/UniswapV2Library.sol (hunk 2 above)
Comment (alex0207s) [reply test/utils/UniswapV2Library.sol:115]: Ok, I see.

**Review by alex0207s — APPROVED** [review]

**Pull request merged**